lib/api_remote.ml

```diff
@@ -113,6 +100,72 @@ let get_permalink ~(ctx : Context.t) ~(req : Slack_t.get_permalink_req) =
     ~name:(sprintf "chat.getPermalink (%s, %s)" req.channel req.message_ts)
     ~ctx `GET api_path Slack_j.read_get_permalink_res
 
+let www_form_of_get_upload_url_ext (req : Slack_t.get_upload_url_ext_req) =
+  let fields =
+    [
+      Some ("filename", req.filename);
+      Some ("length", Int.to_string req.length);
+      string_field_val req.alt_txt "alt_txt";
+      string_field_val req.snippet_type "snippet_type";
+    ]
+  in
+  list_filter_opt fields
+
+let get_upload_url_external ~(ctx : Context.t) ~(req : Slack_t.get_upload_url_ext_req) =
+  log#info "getting upload url for %s" req.filename;
+  let args = www_form_of_get_upload_url_ext req in
+  let data = Web.make_url_args args in
+  let body = `Raw ("application/x-www-form-urlencoded", data) in
+  log#info "data to upload req: %s" data;
+  request_token_auth ~ctx
+    ~name:(sprintf "files.getUploadURLExternal (%s)" req.filename)
+    ~body `POST "files.getUploadURLExternal" Slack_j.read_get_upload_url_ext_res
+
+let complete_upload_external ~(ctx : Context.t) ~(req : Slack_t.complete_upload_ext_req) =
+  log#info "completing upload url for %s" @@ Slack_j.string_of_files_v2 req.files;
+  let data = Slack_j.string_of_complete_upload_ext_req req in
+  let body = `Raw ("application/json", data) in
+  log#info "data to upload req: %s" data;
+  request_token_auth ~ctx
+    ~name:(sprintf "files.completeUploadExternal (%s)" @@ Slack_j.string_of_files_v2 req.files)
+    ~body `POST "files.completeUploadExternal" Slack_j.read_complete_upload_ext_res
+
+(** [upload_file ctx file] upload [file] to channels noted in [file.channels]
+      with content [file.content]; Not supporting file upload through form using
+      `file` currently
+      uses web API with access token *)
+let upload_file ~(ctx : Context.t) ~(file : Slack_t.files_upload_req) =
+  match file.filename, file.content with
+  | None, _ | _, None -> Exn.fail "need to supply both filename and content"
+  | Some filename, Some content ->
+    let length = String.length content in
+    let req = Slack_j.make_get_upload_url_ext_req ~filename ~length () in
+    ( match%lwt get_upload_url_external ~ctx ~req with
+    | Error e -> Lwt.return_error e
+    | Ok { upload_url; file_id; _ } ->
```

Comment:
slack api fails with a 200 OK response but sends the `ok` property of the payload as false and includes an `error` property on the response.

Error responses aren't properly handled here or below, after the `complete_upload_external` response.

Reply:
Yes this is true. However, the api's are calling `slack_api_request` function under the hood that handles this and returns the appropriate thing

https://github.com/ahrefs/slack/blob/365868fc210ce59425e0c458cfd723101980b977/lib/api_remote.ml#L29-L39